src/transformStateWithClones.js

```diff
@@ -1,13 +1,58 @@
 'use strict';
-
 /**
  * @param {Object} state
  * @param {Object[]} actions
  *
  * @return {Object[]}
  */
+
 function transformStateWithClones(state, actions) {
-  // write code here
+  const report = [];
+
+  for (const action of actions) {
+    const stateClone =
+      report.length === 0 ? { ...state } : { ...report[report.length - 1] };
+
+    const actionResult = () => {
+      switch (action.type) {
+        case 'addProperties':
+          return addProperties(stateClone, action.extraData);
+
+        case 'removeProperties':
+          return removeProperties(stateClone, action.keysToRemove);
```

Comment:
The `removeProperties` function modifies the original object by deleting properties. Consider returning a new object with the specified properties removed to avoid side effects.